### PR TITLE
Vulkan: Create renderpass objects on demand.

### PR DESCRIPTION
**Pull request by hrydgard**
Just a little refactoring to prepare for some later features/optimizations. Should have negligible perf impact.

For example, multisampling will require us to support a larger variety of renderpass objects.

Also separates depth and stencil renderpass actions.

### CODE REVIEW

File: ext/native/thin3d/VulkanQueueRunner.cpp [hunk 1]
```diff
@@ -141,13 +141,30 @@ void VulkanQueueRunner::InitBackbufferRenderPass() {
...
-void VulkanQueueRunner::InitRenderpasses() {
+VkRenderPass VulkanQueueRunner::GetRenderPass(VKRRenderPassAction colorAction, VKRRenderPassAction depthAction) {
+	RPKey key{ colorAction, depthAction };
+	auto pass = renderPasses_.Get(key);
+	if (pass) {
+		return pass;
+	}
+
 	// Create a bunch of render pass objects, for normal rendering with a depth buffer,
 	// with clearing, without clearing, and dont-care for both depth/stencil and color, so 3*3=9 combos.
```
Comment (unknownbrackets) [inline ext/native/thin3d/VulkanQueueRunner.cpp]: nit: Outdated comment.

-[Unknown]

File: GPU/Directx9/StencilBufferDX9.cpp [hunk 2]
```diff
@@ -191,7 +191,7 @@ bool FramebufferManagerDX9::NotifyStencilUpload(u32 addr, int size, bool skipZer
 	u16 h = dstBuffer->renderHeight;
 
 	if (dstBuffer->fbo) {
-		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR });
+		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR, Draw::RPAction::CLEAR });
```
Comment (unknownbrackets) [inline GPU/Directx9/StencilBufferDX9.cpp]: * KEEP, KEEP, CLEAR.  Should be supported on 9.

-[Unknown]

File: GPU/D3D11/StencilBufferD3D11.cpp [hunk 3]
```diff
@@ -161,7 +161,7 @@ bool FramebufferManagerD3D11::NotifyStencilUpload(u32 addr, int size, bool skipZ
 	float v1 = 1.0f;
 	MakePixelTexture(src, dstBuffer->format, dstBuffer->fb_stride, dstBuffer->bufferWidth, dstBuffer->bufferHeight, u1, v1);
 	if (dstBuffer->fbo) {
-		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR });
+		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR, Draw::RPAction::CLEAR });
```
Comment (unknownbrackets) [inline GPU/D3D11/StencilBufferD3D11.cpp]: * KEEP, KEEP, CLEAR.  MSDN implies it's possible but I thought there was some reason we were clearing depth here?

-[Unknown]

File: GPU/GLES/StencilBufferGLES.cpp [hunk 4]
```diff
@@ -160,9 +160,9 @@ bool FramebufferManagerGLES::NotifyStencilUpload(u32 addr, int size, bool skipZe
 	Draw::Framebuffer *blitFBO = nullptr;
 	if (useBlit) {
 		blitFBO = GetTempFBO(w, h, Draw::FBO_8888);
-		draw_->BindFramebufferAsRenderTarget(blitFBO, { Draw::RPAction::DONT_CARE, Draw::RPAction::DONT_CARE });
+		draw_->BindFramebufferAsRenderTarget(blitFBO, { Draw::RPAction::DONT_CARE, Draw::RPAction::DONT_CARE, Draw::RPAction::DONT_CARE });
 	} else if (dstBuffer->fbo) {
-		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR });
+		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR, Draw::RPAction::CLEAR });
```
Comment (unknownbrackets) [inline GPU/GLES/StencilBufferGLES.cpp]: * KEEP, KEEP, CLEAR.

-[Unknown]

File: GPU/Vulkan/StencilBufferVulkan.cpp [hunk 5]
```diff
@@ -123,7 +123,7 @@ bool FramebufferManagerVulkan::NotifyStencilUpload(u32 addr, int size, bool skip
 		}
 
 		// TODO: Find a nice way to clear alpha here too.
-		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR });
+		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR, Draw::RPAction::CLEAR });
```
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp]: * KEEP, KEEP, CLEAR.

-[Unknown]

File: GPU/Vulkan/StencilBufferVulkan.cpp [hunk 6]
```diff
@@ -137,7 +137,7 @@ bool FramebufferManagerVulkan::NotifyStencilUpload(u32 addr, int size, bool skip
 	float v1 = 1.0f;
 	MakePixelTexture(src, dstBuffer->format, dstBuffer->fb_stride, dstBuffer->bufferWidth, dstBuffer->bufferHeight, u1, v1);
 	if (dstBuffer->fbo) {
-		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR });
+		draw_->BindFramebufferAsRenderTarget(dstBuffer->fbo, { Draw::RPAction::KEEP, Draw::RPAction::CLEAR, Draw::RPAction::CLEAR });
```
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp]: * KEEP, KEEP, CLEAR.

-[Unknown]

**Review by unknownbrackets — COMMENTED** [review]
Yay, so now we can stop clearing depth when we clear stencil in stencil uploads, right?

-[Unknown]

**Comment by hrydgard** [comment]
I wanted to not actually make the change in this pull request, to avoid any regressions, and make the suggested changes in a followup. Fixing the outdated comment though.

File: GPU/D3D11/StencilBufferD3D11.cpp (hunk 3 above)
Comment (hrydgard) [reply GPU/D3D11/StencilBufferD3D11.cpp]: No idea what that was, probably a misunderstanding... I'll make this change in a followup, want to keep this one regression-free.

Comment (unknownbrackets) [reply GPU/D3D11/StencilBufferD3D11.cpp]: Okay, we did used to only clear stencil on some of these platforms, at least, before BindFramebufferAsRenderTarget.  For some reason I thought this wasn't supported everywhere.

https://github.com/hrydgard/ppsspp/blob/d9acd27126b2f9f7684fdfebbc50d953d1002a77/GPU/GLES/StencilBufferGLES.cpp#L147

https://github.com/hrydgard/ppsspp/blob/d9acd27126b2f9f7684fdfebbc50d953d1002a77/GPU/GLES/StencilBufferGLES.cpp#L197

But yeah, follow up makes sense.

-[Unknown]

Comment (hrydgard) [reply GPU/D3D11/StencilBufferD3D11.cpp]: I also thought so but turns out I was wrong :)

**Comment by unknownbrackets** [comment]
Travis is slow but all the usual suspect platforms are green.

-[Unknown]

**Pull request merged**